### PR TITLE
1.x: Optimizate single just

### PR DESCRIPTION
**Pull request by Chaoba**
This PR add similar optimization as ScalarSynchronousObservable to Single.just method.

### CODE REVIEW

File: src/main/java/rx/Single.java [hunk 1]
```diff
@@ -685,6 +679,9 @@ public void call(SingleSubscriber<? super T> te) {
      * @see <a href="http://reactivex.io/documentation/operators/merge.html">ReactiveX operators documentation: Merge</a>
      */
     public final static <T> Single<T> merge(final Single<? extends Single<? extends T>> source) {
+        if (source.getClass() == ScalarSynchronousSingle.class) {
```
Comment (akarnokd) [inline src/main/java/rx/Single.java]: It appears `instanceof` can be optimized much better by JIT than class equivalence checks. Could you rewrite this and all the other places?

File: src/main/java/rx/internal/util/ScalarSynchronousSingle.java [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+import rx.functions.Func1;
+import rx.internal.schedulers.EventLoopsScheduler;
+
+public final class ScalarSynchronousSingle<T> extends Single<T> {
+
+    public static final <T> ScalarSynchronousSingle<T> create(T t) {
+        return new ScalarSynchronousSingle<T>(t);
+    }
+
+    final T t;
```
Comment (artem-zinnatullin) [inline src/main/java/rx/internal/util/ScalarSynchronousSingle.java]: Let's rename this to `value`? Will improve readability of the code!

**Comment by artem-zinnatullin** [comment]
Nice PR! @Chaoba can you please remove merge commit from the PR (rebase should help)?

**Comment by Chaoba** [comment]
@akarnokd and @artem-zinnatullin  please review on  #3642

**Pull request closed without merging**